File: packages/clinical-contracts/src/rule4/eligibility/eligibilityFingerprintV1.ts
import { createHash } from 'node:crypto';

import { canonicalStableDumps } from '../canonicalJson.js';
import type { Rule4EligibilityAdapterOutput, Rule4SlotEligibilityResolution } from './types.js';
import type { Rule4GateResult } from './types.js';

export const RULE4_CANDIDATE_ELIGIBILITY_FINGERPRINT_V1 =
  'rule4-candidate-eligibility-fingerprint-v1' as const;

function gatePayload(gate: Rule4GateResult): Record<string, unknown> {
  return {
    gate_id: gate.gateId,
    outcome: gate.outcome,
    evidence_item_ids: [...gate.evidenceItemIds].sort(),
    reason_codes: [...gate.reasonCodes].sort(),
    limitation_codes: [...gate.limitationCodes].sort(),
  };
}

function slotPayload(slot: Rule4SlotEligibilityResolution): Record<string, unknown> {
  const gates = [...slot.gateResults]
    .sort((a, b) => a.gateId.localeCompare(b.gateId))
    .map(gatePayload);
  return {
    formula_slot_id: slot.formulaSlotId,
    formula_target_id: slot.formulaTargetId,
    target_role: slot.targetRole,
    eligibility_status: slot.eligibilityStatus,
    candidate_family: slot.candidateFamily,
    eligible_family_options: [...slot.eligibleFamilyOptions].sort(),
    family_gate_status: slot.familyGateStatus,
    gate_results: gates,
    blocking_gate_codes: [...slot.blockingGateCodes].sort(),
    upstream_context_status: slot.upstreamContextStatus,
    selection_status: slot.selectionStatus,
    selected_cascade: slot.selectedCascade,
    selected_dilution: slot.selectedDilution,
    reason_codes: [...slot.reasonCodes].sort(),
    limitation_codes: [...slot.limitationCodes].sort(),
  };
}

export function rule4CandidateEligibilityFingerprintV1Payload(input: {
  rulesetVersion: string;
  registryVersion: string;
  selectionStatus: 'NOT_STARTED';
  slotResolutions: readonly Rule4SlotEligibilityResolution[];
  reasonCodes: readonly string[];
  limitationCodes: readonly string[];
}): string {
  const slots = [...input.slotResolutions]
    .sort((a, b) => a.formulaSlotId.localeCompare(b.formulaSlotId))
    .map(slotPayload);
  return canonicalStableDumps({
    fingerprint_version: RULE4_CANDIDATE_ELIGIBILITY_FINGERPRINT_V1,
    ruleset_version: input.rulesetVersion,
    registry_version: input.registryVersion,
    selection_status: input.selectionStatus,
    slot_resolutions: slots,
    reason_codes: [...new Set(input.reasonCodes)].sort(),
    limitation_codes: [...new Set(input.limitationCodes)].sort(),
  });
}

export function rule4CandidateEligibilityFingerprintV1Hash(
  input: Parameters<typeof rule4CandidateEligibilityFingerprintV1Payload>[0],
): string {
  const payload = rule4CandidateEligibilityFingerprintV1Payload(input);
  return createHash('sha256').update(payload, 'utf8').digest('hex').toUpperCase();
}

export function fingerprintFromEligibilityOutput(
  output: Pick<
    Rule4EligibilityAdapterOutput,
    | 'rulesetVersion'
    | 'registryVersion'
    | 'selectionStatus'
    | 'slotResolutions'
    | 'reasonCodes'
    | 'limitationCodes'
  >,
): string {
  return rule4CandidateEligibilityFingerprintV1Hash({
    rulesetVersion: output.rulesetVersion,
    registryVersion: output.registryVersion,
    selectionStatus: output.selectionStatus,
    slotResolutions: output.slotResolutions,
    reasonCodes: output.reasonCodes,
    limitationCodes: output.limitationCodes,
  });
}

File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/eligibility/eligibility_fingerprint_v1.py
from __future__ import annotations

import hashlib

from ..canonical_json import canonical_stable_dumps

RULE4_CANDIDATE_ELIGIBILITY_FINGERPRINT_V1 = "rule4-candidate-eligibility-fingerprint-v1"


def _gate_payload(gate: dict) -> dict:
    return {
        "gate_id": gate["gate_id"],
        "outcome": gate["outcome"],
        "evidence_item_ids": sorted(gate.get("evidence_item_ids") or []),
        "reason_codes": sorted(gate.get("reason_codes") or []),
        "limitation_codes": sorted(gate.get("limitation_codes") or []),
    }


def _slot_payload(slot: dict) -> dict:
    gates = sorted(slot.get("gate_results") or [], key=lambda g: g["gate_id"])
    return {
        "formula_slot_id": slot["formula_slot_id"],
        "formula_target_id": slot.get("formula_target_id"),
        "target_role": slot["target_role"],
        "eligibility_status": slot["eligibility_status"],
        "candidate_family": slot["candidate_family"],
        "eligible_family_options": sorted(slot.get("eligible_family_options") or []),
        "family_gate_status": slot["family_gate_status"],
        "gate_results": [_gate_payload(g) for g in gates],
        "blocking_gate_codes": sorted(slot.get("blocking_gate_codes") or []),
        "upstream_context_status": slot["upstream_context_status"],
        "selection_status": slot["selection_status"],
        "selected_cascade": slot.get("selected_cascade"),
        "selected_dilution": slot.get("selected_dilution"),
        "reason_codes": sorted(slot.get("reason_codes") or []),
        "limitation_codes": sorted(slot.get("limitation_codes") or []),
    }


def rule4_candidate_eligibility_fingerprint_v1_payload(
    *,
    ruleset_version: str,
    registry_version: str,
    selection_status: str,
    slot_resolutions: list[dict],
    reason_codes: list[str],
    limitation_codes: list[str],
) -> str:
    slots = sorted(slot_resolutions, key=lambda s: s["formula_slot_id"])
    return canonical_stable_dumps(
        {
            "fingerprint_version": RULE4_CANDIDATE_ELIGIBILITY_FINGERPRINT_V1,
            "ruleset_version": ruleset_version,
            "registry_version": registry_version,
            "selection_status": selection_status,
            "slot_resolutions": [_slot_payload(s) for s in slots],
            "reason_codes": sorted(set(reason_codes)),
            "limitation_codes": sorted(set(limitation_codes)),
        }
    )


def rule4_candidate_eligibility_fingerprint_v1_hash(
    *,
    ruleset_version: str,
    registry_version: str,
    selection_status: str,
    slot_resolutions: list[dict],
    reason_codes: list[str],
    limitation_codes: list[str],
) -> str:
    payload = rule4_candidate_eligibility_fingerprint_v1_payload(
        ruleset_version=ruleset_version,
        registry_version=registry_version,
        selection_status=selection_status,
        slot_resolutions=slot_resolutions,
        reason_codes=reason_codes,
        limitation_codes=limitation_codes,
    )
    return hashlib.sha256(payload.encode("utf-8")).hexdigest().upper()


def fingerprint_from_eligibility_output(output: dict) -> str:
    return rule4_candidate_eligibility_fingerprint_v1_hash(
        ruleset_version=output["ruleset_version"],
        registry_version=output["registry_version"],
        selection_status=output["selection_status"],
        slot_resolutions=list(output["slot_resolutions"]),
        reason_codes=list(output["reason_codes"]),
        limitation_codes=list(output["limitation_codes"]),
    )
